webhooks/jira: Ignore payloads without an explicit event type.

eeshangarg opened the pull request:
We recently ran into a payload in production that didn't contain
an event type at all. A payload where we can't figure out the event
type is by definition not a webhook payload, so we should ignore
it.

@timabbott FYI :)

<!-- What's this PR for?  (Just a link to an issue is fine.) -->

**Testing plan:** <!-- How have you tested? -->

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on zerver/webhooks/jira/view.py:
```diff
@@ -346,7 +346,7 @@ def api_jira_webhook(
     event = get_event_type(payload)
-    if event in IGNORED_EVENTS:
+    if event is None or event in IGNORED_EVENTS:
         return json_success()
```
I don't think this is the right fix; if we wanted to make a silent success in this case, we should remove the `event is not None` check just after.  But I think we want this to be logged as unsupported.  The right fix would probably instead to do `content_func = None` on the first line of this function, but we need to pass an `event` to that function.  

We might also want to figure out why `get_event_type` is returning `None` here.  (But also we should probably raise some sort of exception for an event we can't parse).

eeshangarg commented:
@timabbott This PR has now been updated and is ready for another review. Thanks! :)

timabbott commented on zerver/decorator.py:
```diff
@@ -693,11 +694,11 @@ def _wrapped_func_arguments(
                     raise err
 
-                if isinstance(err, UnsupportedWebhookEventType):
+                if isinstance(err, (UnsupportedWebhookEventType, AnomalousWebhookPayload)):
```
I don't think these changes are correct.  The intent is that `UnsupportedWebhookEventType` should not be returned to users as a 400 error, because the request from the third party is a valid event -- we just don't do anything for that event. 

Whereas for `AnomalousWebhookPayload`, it's a 400 error -- we think that you're sending requests from a thing that is not a Jira server to Zulip.

timabbott replied:
I'm not sure how to do the plumbing to have `webhook_name` get set if we remove this; might require some fiddling.

timabbott replied:
From a logging perspective, I think we want these to go to `webhooks_unexpected_events.log`, whereas the `UnsupportedWebhookEventType` should go to `webhooks_unsupported_events.log`.

```
webhooks_unexpected_events.log
webhooks_unsupported_events.log
```

eeshangarg replied:
@timabbott As for your first comment, isn't `UnsupportedWebhookEventType` already being returned as a 400 error since the exception subclasses `JsonableError` with a default status code of 400? Note that `UnsupportedWebhookEventType`, when caught, is re-raised at the end after logging. Are you saying that this PR should make it such that `UnsupportedWebhookEventType` should not be re-raised after logging?

Moreover, I would say that both `AnomalousWebhookPayload` and `UnsupportedWebhookEventType` represent a 400 error in the sense that both are "bad requests" that we can't handle, either because we choose not to, or because the payload seems incorrectly formatted or from an incorrect service. Conceptually, even `UnsupportedWebhookEventType` should be presented as an error that the user can report so that if we get too many of these for one event type, we can add support for said event.

As for your comment about the log file name, I named it `webhooks_anomalous_payloads.log`, since the scope of the term `anomalous payloads` is wider than that for the term `unexpected events`, which could be misinterpreted as just being about service-specific webhook events, and not just about anomalous payloads in general. Thanks!

alexmv replied:
I like `webhooks_anomalous_payloads.log` as a name, both because it matches the exception class, and also because it's harder than `unexpected` to confuse for `unsupported` at a glance. :)

alexmv replied:
On the HTTP status code front, we are certainly returning 400's currently for `UnsupportedWebhookEventType` exceptions.  I think there's a potential argument to make those 501's, but [the specs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) imply that this is generally treated as the _HTTP method_ being unimplemented, not the resource.

eeshangarg replied:
501s don't seem appropriate here either, I agree. Honestly, I think 400s are fine for both of these, these are both classes of errors that should fall under the same larger category of stuff that can go wrong with our webhooks. Even if we choose to not support an event, that can still be an error in the sense that maybe we should if there are lots of these exceptions for the same type of payload/event.

timabbott commented:
I pushed a commit adjusting the strings, and also posted a comment where I think it's a bug to not return the exception to the end user here.

timabbott commented:
@alexmv FYI since in some sense the Sentry setup is more important than the log file format, but I'd like confirmation on my thinking here.

eeshangarg commented:
@timabbott I just updated this PR (squashed and rebased as well). I also left a comment above in response to some of your feedback. This PR is ready for another review, thanks! :)

alexmv commented on zerver/decorator.py:
```diff
@@ -308,8 +310,11 @@ def access_user_by_api_key(
 def log_exception_to_webhook_logger(
     summary: str,
     unsupported_event: bool,
+    anomalous_payload: Optional[bool] = False,
 ) -> None:
-    if unsupported_event:
+    if anomalous_payload:
+        webhook_anomalous_payloads_logger.exception(summary, stack_info=True)
+    elif unsupported_event:
         webhook_unsupported_events_logger.exception(summary, stack_info=True)
     else:
         webhook_logger.exception(summary, stack_info=True)
```
I'm not thrilled with what is really a tri-state value being expressed as two `bool`s.  Can we just pass down the `err` exception object, rather than its `str` and a bunch of bools?

eeshangarg replied:
Good point! Thanks! :)

alexmv commented on zerver/decorator.py:
```diff
@@ -358,15 +363,16 @@ def _wrapped_func_arguments(
 
                     notify_bot_owner_about_invalid_json(user_profile, webhook_client_name)
                 elif isinstance(err, JsonableError) and not isinstance(
-                    err, UnsupportedWebhookEventType
+                    err, (UnsupportedWebhookEventType, AnomalousWebhookPayload)
                 ):
                     pass
                 else:
-                    if isinstance(err, UnsupportedWebhookEventType):
+                    if isinstance(err, (UnsupportedWebhookEventType, AnomalousWebhookPayload)):
                         err.webhook_name = webhook_client_name
                     log_exception_to_webhook_logger(
                         summary=str(err),
                         unsupported_event=isinstance(err, UnsupportedWebhookEventType),
+                        anomalous_payload=isinstance(err, AnomalousWebhookPayload),
```
This sort of pattern really makes me wonder if `AnomalousWebhookPayload` should be a `UnsupportedWebhookEventType` subclass, or both derive from a common root class or protocol.  I don't have any namings that pop to mind for it immediately, though.

eeshangarg replied:
That's a great idea, would certainly simplify the code that catches these webhook exceptions.

alexmv commented:
Looks reasonable to me.  /cc @andersk who has recently been looking into better general systems for validating webhook inputs and failing out when the inputs don't contain what they should.

timabbott commented:
Merged, thanks @eeshangarg!  We may have further improvements we want to do here, but it seems likely they can happily build on top of our having created this second class of webhook exception.
